### PR TITLE
Hotfix/stand alone eigensolver

### PR DESCRIPTION
**Pull request by cpviolator**
This PR adds gtest functionality to the Wilson type eigensolvers. The test will compute the following variations:

1. Normal operator and Direct operator
2. Preconditioned and unpreconditioned
3. With SVD if the normal operator is chosen
4. All possible spectrum types available to the eigensolver.
5. Thick Restarted Lanczos, Block Thick Restarted Lanczos, Implicitly Restarted Arnoldi.

Each test will be performed with no polynomial acceleration as that requires the user to fine tune input parameters tailored to a specific input matrix.

For the SVD testing, the routine will compute the L_2 norm of
```
M * X_left - sigma * X_right.
```
This explicitly tests the SVD data, and implicitly tests the underlying eigenvector data, as the SVD of M(Mdag) is computed via the eigendecomposition of MdagM(MMdag). For the Eigenvectors produced in Direct solves, the host will measure the L_2 norm of
```
M * X - lambda * X.
```

This gtest was tested with Wilson, Clover, Twisted-Mass, Twisted Clover, and all Domain Wall actions. For the Block solver, a block size of 4 was used, and a batched rotation size of 4 was used for all tests. The following command can be used to test this PR:

```
#!/bin/bash

export QUDA_RESOURCE_PATH=.

for ACTION in wilson clover twisted-mass twisted-clover domain-wall domain-wall-4d mobius mobius-eofa ; do
	      
    ./eigensolve_test --sdim 4 --tdim 4 --gridsize 1 1 1 1 --eig-n-conv 12 --eig-n-ev 24 --eig-n-kr 96 --eig-tol 1e-6 --mass 0.01 --eig-max-restarts 1000 --verbosity verbose --prec single --enable-testing true --verify true --dslash-type ${ACTION}
done
```

### CODE REVIEW

**Comment by kostrzewa** [comment]
@cpviolator is the standalone eigensolver capable of computing eigenvalues also for the non-degenerate twisted mass (clover) operators? The reason that I ask is that we use RHMC to simulate the non-degenerate doublet and it is necessary to check in the simulation that our eigenvalue bounds are appropriate for the chosen parameters. It's on my TODO list to add to enable our QUDA interface to perform the necessary measurement of the minimal and maximal evals using QUDA.

**Comment by cpviolator** [comment]
@kostrzewa I've yet to come across an action that the eigensolver cannot handle. Some dslash options such as laplace or clover hasenbuch have no host side verification routines, and I see also now that non-degenerate twisted (clover) have no explicit command lined dslash option, one must pass that as an extra command. I'm confident that the eigensolver can handle it. Allow me to make the extra modifications to test this path and I'll get back to you.

File: lib/interface_quda.cpp
Comment (maddyscientist) [inline lib/interface_quda.cpp]: If one is interfacing with an application that requires UKQCD ordering or another ordering completely, hardcoding this to DeGrand-Rossi doesn't seem like the right approach.  Shouldn't it just use what value of `gamma_basis` the user has passed in `QudaEigParam::QudaInvertParam`?

File: lib/interface_quda.cpp
Comment (maddyscientist) [inline lib/interface_quda.cpp]: Probably delete the `(I think...)`

File: tests/eigensolve_test_gtest.hpp
Comment (maddyscientist) [inline tests/eigensolve_test_gtest.hpp]: What's this factor of 5 for?

File: lib/interface_quda.cpp
Comment (maddyscientist) [inline lib/interface_quda.cpp]: typo (gaige)

File: lib/interface_quda.cpp
Comment (maddyscientist) [inline lib/interface_quda.cpp:2281]: While reading this, I was thinking to suggest modifying this to vector of fields, rather than vector of objects.  But on second though, I think it makes more sense to get this PR in with all the tests and only then to perform surgery of that nature.

**Review by maddyscientist — COMMENTED** [review]
@cpviolator this is an excellent PR, thank you for adding all these tests.  Beyond the minor comments I made, the main remaining ask is to add test invocations to tests/CMakeLists.txt in a similar fashion as is already done for the solver tests.

**Comment by cpviolator** [comment]
@kostrzewa I checked the non-deg twisted (clover) paths with the eigensolver and all seems well. I will caveat that by saying that when I tried to run with `QUDA_OPENMP=ON` I saw some regression, so I would avoid that option for the moment.

As soon as this PR is complete please feel free to reach out so we can discuss proper command line/param options for the eigendecompostion of non-deg operators.

File: lib/interface_quda.cpp
Comment (cpviolator) [reply lib/interface_quda.cpp:2222]: All works well without this, so removed. 7e005b04f

Comment (cpviolator) [reply lib/interface_quda.cpp:2222]: 7e005b04f

Comment (cpviolator) [reply lib/interface_quda.cpp]: All works well without this, so removed. 7e005b04f

File: lib/interface_quda.cpp
Comment (cpviolator) [reply lib/interface_quda.cpp]: 7e005b04f

File: tests/eigensolve_test_gtest.hpp
Comment (cpviolator) [reply tests/eigensolve_test_gtest.hpp]: The IRAM eigensolver will sometimes report convergence with tolerances slightly higher than requested. The same phenomenon occurs in ARPACK. This fudge factor prevents failure when IRAM has solved to say 2e-6 when 1e-6 is requested. The solution to avoid this is to use a Krylov space (n-kr) about 3-4 times the size of the search space (n-ev), or use a well chosen Chebyshev polynomial, or use a tighter than necessary tolerance.

File: lib/interface_quda.cpp
Comment (cpviolator) [reply lib/interface_quda.cpp]: https://github.com/lattice/quda/commit/7e005b04fb5aeeec124c87bfb099a9b4580591e9

File: tests/eigensolve_test_gtest.hpp
Comment (maddyscientist) [reply tests/eigensolve_test_gtest.hpp]: Ok, could you leave a comment to this effect then and only apply the fudge factor if doing IRAM?

File: lib/eigensolve_quda.cpp
Comment (weinbe2) [inline lib/eigensolve_quda.cpp:143]: I thought that a point of the SVD was that it exists for any matrix regardless of its property and shape? https://en.wikipedia.org/wiki/Singular_value_decomposition

Comment (cpviolator) [reply lib/eigensolve_quda.cpp:143]: @weinbe2 Yes, that's right, but I haven't coded up a routine that can compute the SVD of a non-hermitain matrix M. Instead, we compute the eigendecomposition of MdagM, which also furnishes the right singular vectors of M:
```
MdagM * X = lambda X
MdagM * X_right = Mdag * sigma * X_left = sigma^2 X_right
``` 
hence the sigma and X_right data can be computed via the (admittedly poorer conditioned) MdagM operator. The X_left vectors are easily identified via
```
M * X_right = sigma * X_left
```
I think it's OK to use the poorer MdagM as we only use the SVD in QUDA to deflate non-hermitian solvers, so the extra time required to identify the SVD via MdagM rather than M is amortised by the solver speed-up.

File: tests/eigensolve_test_gtest.hpp
Comment (cpviolator) [reply tests/eigensolve_test_gtest.hpp]: Done 9b58a9770

**Comment by cpviolator** [comment]
@lattice/quda_core I'll Give this one more full build with SANITIZE on and roll through all the actions one last time and dump the stdout here in a .txt file for reference, as well as adding test invocations to tests/CMakeLists.txt.

**Comment by cpviolator** [comment]
@kostrzewa I added the ndeg twisted(clover) tests to the ctest suite and all is working well. If you have any difficulty obtaining a spectrum let me know and I'll help tweak the Arnoldi parameters.

**Review by weinbe2 — APPROVED** [review]
I just gave this a quick spin and everything looks good from the staggered side, and I didn't have to change any flags for coarsest-level deflation in MG to go through which is great.. I was a bit confused about explicitly adding a flag for the preconditioned vs non-preconditioned operator, but after thinking about it/toying with test runs a bit I came to appreciate its clarity. Thanks for the contribution @cpviolator !

This has my approval, but I think it'd be good if @kostrzewa or another applications person gave the PR a final spin before hitting merge. Happy to be overriden, though.

**Comment by kostrzewa** [comment]
Please go ahead with the merge. I won't really have the time to work on the interface for our non-degenerate operator before September or so... @cpviolator many thanks for taking nd twisted mass (clover) into account: this will prove very useful in the near future

**Comment by weinbe2** [comment]
Just posted in with @maddyscientist and there are still a few more tests she wants to do, so no hitting merge just yet.

**Review by maddyscientist — APPROVED** [review]
Approving.  The one outstanding thing I wanted to do was fix the function `display_test_info` so that each invoked test's parameters would be printed out correctly, instead of just printing out the initial values repeatedly.

(I also sneaked in a link error with quarter precision that cropped up elsewhere.)

I think we can merge this when Jenkins is done.

**Pull request merged**
